verus/lcf/src/kernel.rs

```diff
@@ -0,0 +1,1108 @@
+    //aids verus verifier in reasoning about equality; not fully sufficient - must call prop_eq for certain types
+    fn eq(&self, other: &Self) -> (res: bool)
+        ensures
+            res <==> self.deep_view() == other.deep_view(),
+    {
+        match (self, other) {
+            (Prop::App(s, v), Prop::App(t, w)) => s == t && v == w,
+            (Prop::Eq(a, b), Prop::Eq(c, d)) => a == c && b == d,
+            (Prop::App(_, _), Prop::Eq(_, _)) | (Prop::Eq(_, _), Prop::App(_, _)) => false,
+        }
+    }
+}
+
+impl SpecProp {
+    //checks if Prop variants contain key to map; calls spec_complete_subst from Term
+    pub open spec fn spec_complete_subst(self, s: SpecSubst) -> bool {
+        match self {
+            SpecProp::App(head, args) => forall|i: int|
+                #![auto]
+            //forall checks every index in list of SpecTerms in args to check if they are complete substitutions
+
+                0 <= i < args.len() ==> args[i].spec_complete_subst(s),
+            SpecProp::Eq(x, y) => x.spec_complete_subst(s) && y.spec_complete_subst(s),
+        }
+    }
+
+    //checks if Prop variants are of base types
+    pub open spec fn spec_concrete(self) -> bool {
+        match self {
+            SpecProp::App(head, args) => forall|i: int|
+                #![auto]
+            //forall checks every index in list of SpecTerms in args to check if they are concrete
+
+                0 <= i < args.len() ==> args[i].spec_concrete(),
+            SpecProp::Eq(x, y) => x.spec_concrete() && y.spec_concrete(),
+        }
+    }
+
+    //checks if SpecProp is an App variant
+    pub open spec fn spec_symbolic(self) -> bool {
+        self is App
+    }
+
+    //checks whether valid SpecProp types are concrete and are equal
+    pub open spec fn spec_valid(self) -> bool {
+        if self.spec_symbolic() || !(self.spec_concrete()) {
+            false
+        } else {
+            match self {
+                SpecProp::Eq(x, y) => match (x, y) {
+                    (SpecTerm::Const(x), SpecTerm::Const(y)) => x == y,
+                    (SpecTerm::Var(x), SpecTerm::Var(y)) => false,
+                    (SpecTerm::Const(_), SpecTerm::Var(_))
+                    | (SpecTerm::Var(_), SpecTerm::Const(_)) => false,
+                },
+                SpecProp::App(s, v) => false,
+            }
+        }
+    }
+
+    //reasons about substitution of SpecProp variants to SpecTerms
+    pub open spec fn spec_subst(self, s: SpecSubst) -> (res: SpecProp)
+        recommends
+            self.spec_complete_subst(s),
+    {
+        match self {
+            SpecProp::App(h, args) => {
+                let new_sequence = args.map_values(|p: SpecTerm| p.spec_subst(s));
+                SpecProp::App(h, new_sequence)
+            },
+            SpecProp::Eq(x, y) => SpecProp::Eq(x.spec_subst(s), y.spec_subst(s)),
+        }
+    }
+}
+
+impl Prop {
+    //proves equality for Prop types. uses term_eq and terms_eq
```

Comment:
I wouldn't say `proves` here---it's an executable function that **checks** if two `Prop`s are equal. Best to be careful about the use of the word `proof` when doing verification!